objects/src/notes/note_execution_hint.rs

```diff
@@ -0,0 +1,182 @@
+    pub fn after_block(block_num: u32) -> Self {
+        NoteExecutionHint::AfterBlock { block_num }
+    }
+
+    /// Creates a [NoteExecutionHint::OnBlockSlot] for the given parameters
+    pub fn on_block_slot(epoch_len: u8, slot_len: u8, slot_offset: u8) -> Self {
+        NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset }
+    }
+
+    /// Returns whether the note execution conditions validate for the given `block_num`
+    pub fn can_be_consumed(&self, block_num: u32) -> bool {
+        match self {
+            NoteExecutionHint::None => false,
+            NoteExecutionHint::Always => true,
+            NoteExecutionHint::AfterBlock { block_num: hint_block_num } => {
+                block_num >= *hint_block_num
+            },
+            NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset } => {
+                let epoch_len_blocks: u32 = 1 << epoch_len;
+                let slot_len_blocks: u32 = 1 << slot_len;
+
+                let block_epoch_index = block_num / epoch_len_blocks;
+
+                let slot_start_block =
+                    block_epoch_index * epoch_len_blocks + (*slot_offset as u32) * slot_len_blocks;
+                let slot_end_block = slot_start_block + slot_len_blocks;
+
+                block_num >= slot_start_block && block_num < slot_end_block
+            },
+        }
+    }
+}
+
+// CONVERSIONS
+// ================================================================================================
+
+impl From<NoteExecutionHint> for u64 {
+    fn from(hint: NoteExecutionHint) -> Self {
+        match hint {
+            NoteExecutionHint::None => 0,
+            NoteExecutionHint::Always => 1 << 32,
+            NoteExecutionHint::AfterBlock { block_num } => {
+                let tag = 2;
+                (tag << 32) | (block_num as u64)
+            },
+            NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset } => {
+                let tag: u64 = 3;
+                let payload: u64 = (((epoch_len as u32) << 16)
+                    | ((slot_len as u32) << 8)
+                    | (slot_offset as u32)) as u64;
+                (tag << 32) | payload
+            },
+        }
+    }
+}
+
+impl TryFrom<u64> for NoteExecutionHint {
+    type Error = NoteError;
+    fn try_from(encoded: u64) -> Result<Self, NoteError> {
+        let tag = ((encoded >> 32) & 0xFF) as u8;
+        let payload = (encoded & 0xFFFFFF) as u32;
+
+        match tag {
+            0 => Ok(NoteExecutionHint::None),
+            1 => Ok(NoteExecutionHint::Always),
+            2 => Ok(NoteExecutionHint::AfterBlock { block_num: payload }),
+            3 => {
+                let epoch_len = ((payload >> 16) & 0xFF) as u8;
+                let slot_len = ((payload >> 8) & 0xFF) as u8;
+                let slot_offset = (payload & 0xFF) as u8;
+                let hint = NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset };
```

Comment:
Similar to the last comment: we don't seem to check here if the 4th byte is valid. It may make sense to ensure that it is always set to 0.

Reply:
Added validations for these cases, and also changed the conversions to be based on `(tag, payload)`